modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdSelectivity.java

```diff
@@ -110,6 +122,654 @@ public Double getSelectivity(IgniteSortedIndexSpool rel, RelMetadataQuery mq, Re
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        return estimateRangeSelectivity(colStat, pred);
+    }
+
+    /**
+     * Get column statistics.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable to get statistics from.
+     * @param pred Predicate to get statistics by related column.
+     * @return ColumnStatistics or {@code null}.
+     */
+    private ColumnStatistics getColStat(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexNode pred) {
+        SqlKind predKind = pred.getKind();
+
+        if (predKind != SqlKind.IS_NULL && predKind != SqlKind.IS_NOT_NULL && predKind != SqlKind.LOCAL_REF &&
+            predKind != SqlKind.NOT && !predKind.belongsTo(SqlKind.COMPARISON))
+            return null;
+
+        RexSlot operand = getOperand(pred, RexSlot.class);
+
+        if (operand == null)
+            return null;
+
+        return getColStatBySlot(rel, mq, tbl, operand);
+    }
+
+    /**
+     * Get column statistics.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable to get statistics from.
+     * @param pred RelSlot to get statistics by related column.
+     * @return ColumnStatistics or {@code null}.
+     */
+    private ColumnStatistics getColStatBySlot(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexSlot pred) {
+        Set<RelColumnOrigin> origins = null;
+
+        if (pred instanceof RexLocalRef) {
+            if (rel instanceof ProjectableFilterableTableScan) {
+                origins = Collections.singleton(
+                    ((ProjectableFilterableTableScan)rel).columnOriginsByRelLocalRef(pred.getIndex()));
+            }
+        }
+        else
+            origins = mq.getColumnOrigins(rel, pred.getIndex());
+
+        if (origins == null || origins.isEmpty() || origins.iterator().next().isDerived())
+            return null;
+
+        IgniteTypeFactory typeFactory = Commons.typeFactory(rel);
+
+        List<String> columns = tbl.getRowType(typeFactory).getFieldNames();
+
+        String colName = columns.get(origins.iterator().next().getOriginColumnOrdinal());
+
+        if (QueryUtils.KEY_FIELD_NAME.equals(colName))
+            colName = tbl.descriptor().typeDescription().keyFieldName();
+
+        Statistic stat = tbl.getStatistic();
+
+        if (stat == null)
+            return null;
+
+        return ((IgniteStatisticsImpl)stat).getColumnStatistics(colName);
+    }
+
+    /**
+     * Estimate range selectivity based on predicate.
+     *
+     * @param colStat Column statistics to use.
+     * @param pred  Condition.
+     * @return Selectivity.
+     */
+    private double estimateRangeSelectivity(ColumnStatistics colStat, RexNode pred) {
+        if (pred instanceof RexCall) {
+            RexLiteral literal = getOperand(pred, RexLiteral.class);
+
+            if (literal == null)
+                return 1.;
```

Comment:
What about other types of operands? It's true that we cannot compute the selectivity properly, but we cannot just omit such predicates as well. In my opinion, predicate `timestamp >= now() - interval '1 day'` should result a better selectivity than no predicate at all.

Reply:
The same for `estimateEqualsSelectivity`